js/IdentityProxyDeploy.js

```diff
@@ -1,15 +1,87 @@
+const swarmhash = require('swarmhash')
 const abi = require('ethereumjs-abi')
 const keccak256 = require('js-sha3').keccak256
 const assert = require('assert')
 
-function getMappingSstore(slotNumber, keyType, key, value) {
+function Generator() {
+	this.jumpdest = new Map();
+	this.tpl = '';
+	return this
+}
+
+Generator.prototype.append = function (code) {
+	this.tpl += code;
+	return this;
+}
+
+Generator.prototype.appendDest = function (tag, code) {
+	this.jumpdest.set(tag, this.tpl);
+	this.tpl += code;
+	return this;
+}
+
+Generator.prototype.addSstore = function (slotNumber, keyType, key, value) {
 	// https://blog.zeppelin.solutions/ethereum-in-depth-part-2-6339cf6bddb9
 	const buf = abi.rawEncode([keyType, 'uint256'], [key, slotNumber])
 	const slot = keccak256(buf)
+	this.append(`${Generator.pushx(value)}${Generator.pushx(slot)}55`);
 	return `sstore(0x${slot}, ${value})`
 }
 
+Generator.prototype.assemble = function (dataSize, subTagSize) {
+	let bytecode = this.tpl;
+	let tagSize = this.tagSize(dataSize, subTagSize);
+	this.jumpdest.forEach((dest, tag) => {
+		const pos = tagSize === 1 ? dest.length / 2 : dest.replace(new RegExp('t', 'g'), 'tag').length / 2;
+		bytecode = bytecode.replace(new RegExp(tag, 'g'), Generator.pushx(pos, tagSize));
+	});
+	return bytecode;
+}
+
+Generator.prototype.pos = function (tag) {
+	const tagSize = this.tagSize();
+	const dest = this.jumpdest.get(tag);
+	if (dest) {
+		return tagSize === 1 ? dest.length / 2 : dest.replace(new RegExp('t', 'g'), 'tag').length / 2;
+	}
+	return -1;
+}
+
+Generator.prototype.tagSize = function (dataSize, subTagSize) {
+	let tagSize = (dataSize ? this.tpl.length + dataSize : this.tpl.length) / 2 > 255 ? 2 : 1;
+	return (subTagSize && subTagSize > tagSize) ? subTagSize : tagSize;
+}
+
+Generator.genMetadataHashBytecode = function (content) {
+	const srcKeccakHash = keccak256(content);
+	const srcSwarmHash = swarmhash(Buffer.from(content)).toString('hex');
+	const metadata = `{"compiler":{"version":"0.5.6+commit.b259423e"},"language":"Solidity","output":{"abi":[{"inputs":[],"payable":false,"stateMutability":"nonpayable","type":"constructor"},{"payable":false,"stateMutability":"nonpayable","type":"fallback"}],"devdoc":{"methods":{}},"userdoc":{"methods":{}}},"settings":{"compilationTarget":{"browser/IdentityProxy.sol":"IdentityProxy"},"evmVersion":"petersburg","libraries":{},"optimizer":{"enabled":true,"runs":200},"remappings":[]},"sources":{"browser/IdentityProxy.sol":{"keccak256":"0x${srcKeccakHash}","urls":["bzzr://${srcSwarmHash}"]}},"version":1}`;
+	const metadataHash = swarmhash(Buffer.from(metadata)).toString('hex');
+	// https://solidity.readthedocs.io/en/v0.5.6/metadata.html#encoding-of-the-metadata-hash-in-the-bytecode
+	// 0xa1 0x65 'b' 'z' 'z' 'r' '0' 0x58 0x20 <32 bytes swarm hash> 0x00 0x29
+	const metadataHashBytecode = `a165627a7a72305820${metadataHash}0029`;
+	return metadataHashBytecode;
+}
+
+Generator.pushx = function (value, minlen) {
+	let rs;
```

Comment:
maybe this could be easier if you just take the input hex, construct a Buffer/arrayish, and then serialize that buffer back

eliminates need to handle edge cases like, checking for valid hex, padding with 0 if length is odd, etc.

Reply:
as i know it's still necessary to check odd length and check valid hex when using Buffer or Array, but maybe i missed something?